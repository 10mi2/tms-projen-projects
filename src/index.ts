export * from "./tms-typescript-app-project";
export * from "./tms-nestjs-app-project";
