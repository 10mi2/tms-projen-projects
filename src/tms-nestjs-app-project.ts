import * as path from "node:path";
import { SampleDir, SampleFile, javascript } from "projen";
import { deepMerge } from "projen/lib/util";
import {
  TmsTypeScriptAppProject,
  TmsTypeScriptAppProjectOptions,
} from "./tms-typescript-app-project";

export interface TmsNestJSAppProjectOptions
  extends TmsTypeScriptAppProjectOptions {
  /**
   * Which type of sample code to include, if `sampleCode` is true.
   *
   * @default "graphql-codefirst"
   * @featured
   */
  readonly sampleType?: "graphql-codefirst" | "graphql-schemafirst" | "rest";
}

/**
 * Create a [NestJS](https://docs.nestjs.com) TypeScript project with [Ten Mile Square](https://tenmilesquare.com)
 * opinionated defaults, and preloaded with one of a few possible sample applications.
 *
 * @pjid tms-nestjs-app
 */

export class TmsNestJSAppProject extends TmsTypeScriptAppProject {
  constructor(options: TmsNestJSAppProjectOptions) {
    const defaultOptions = {
      addDefaultBundle: false,

      esmSupportConfig: false,

      tsconfig: {
        compilerOptions: {
          module: "node16",

          // Needed for nestjs
          noImplicitAny: true,
          declaration: true,
          emitDecoratorMetadata: true,
          experimentalDecorators: true,
          allowSyntheticDefaultImports: true,
          sourceMap: true,
          outDir: "./dist",
          baseUrl: "./",

          // Let the "extends" handle these
          target: undefined,
          forceConsistentCasingInFileNames: undefined,
          strictNullChecks: undefined, // Default to true when strict: true
          noEmit: undefined,
          noFallthroughCasesInSwitch: undefined,

          // Handled below with overrides
          // removeComments: true,
          // incremental: true,
          // strictBindCallApply: true,
        },
      },

      tsconfigDev: {
        compilerOptions: {
          baseUrl: "./",

          target: "es2022",
          experimentalDecorators: undefined,
          emitDecoratorMetadata: undefined,
          allowSyntheticDefaultImports: undefined,
          outDir: undefined,
          skipLibCheck: undefined,
          strictNullChecks: undefined,
          noImplicitAny: undefined,
          forceConsistentCasingInFileNames: undefined,
          noFallthroughCasesInSwitch: undefined,
          noEmit: true,
        },
      },

      jestOptions: {
        jestConfig: {
          // testEnvironment: "node",
          moduleFileExtensions: ["js", "json", "ts"],
          rootDir: ".",
          transform: {
            "^.+\\.(t|j)s$": new javascript.Transform("ts-jest", {
              useESM: false,
              tsconfig: "tsconfig.dev.json",
            }),
          },
          // collectCoverageFrom: ["**/*.(t|j)s"],
          // coverageDirectory: "../coverage",
          testMatch: ["<rootDir>/(test|src)/**/*(*)@(.|-)@(spec|test).ts?(x)"],
          moduleNameMapper: {
            "^(\\.{1,2}/.*)\\.js$": "$1",
          },
        },
        updateSnapshot: javascript.UpdateSnapshot.NEVER,
      },

      sampleCode: true,
      sampleType: "graphql-codefirst",
    } satisfies Partial<TmsNestJSAppProjectOptions>;

    const mergedOptions = deepMerge(
      [
        defaultOptions, // will get mutated
        options,
      ],
      true,
    ) as TmsNestJSAppProjectOptions; // justified cast, since we know we're merging two of this type

    super({ ...mergedOptions, sampleCode: false });

    this.compileTask.reset("nest build");
    if (this.eslint) {
      this.addTask("format").spawn(this.eslint.eslintTask);
    }
    this.addTask("start").exec("nest start");
    this.addTask("start:dev").exec("nest start --watch");
    this.addTask("start:debug").exec("nest start --debug --watch");
    this.addTask("start:prod").exec("node dist/main");

    this.addDeps(
      "@nestjs/common@^10.0.0",
      "@nestjs/core@^10.0.0",
      "@nestjs/platform-express@^10.0.0",
      "reflect-metadata@^0.1.13",
      "rxjs@^7.8.1",

      // graphql
      "graphql",
      "@graphql-tools/utils",
      "graphql-query-complexity",
      "class-transformer",
      "class-validator",
      "graphql-subscriptions",
      "@apollo/server",
      "@nestjs/graphql",
      "@nestjs/apollo",
    );
    this.addDevDeps(
      "@tsconfig/node18",
      "@tsconfig/strictest",
      "@nestjs/cli@^10.0.0",
      "@nestjs/schematics@^10.0.0",
      "@nestjs/testing@^10.0.0",
      "@types/express@^4.17.17",
      "@types/supertest@^2.0.12",
      "supertest@^6.3.3",
      "source-map-support@^0.5.21",
      "ts-loader@^9.4.3",
      "tsconfig-paths@^4.2.0",
    );

    if (this.tsconfig) {
      // Options not yet internally supported by projen
      this.tsconfig.file.addOverride("compilerOptions.removeComments", true);
      this.tsconfig.file.addOverride("compilerOptions.incremental", true);
      this.tsconfig.file.addOverride(
        "compilerOptions.strictBindCallApply",
        true,
      );
    }

    if (mergedOptions.sampleCode) {
      if (mergedOptions.sampleType === "graphql-codefirst") {
        new SampleDir(this, this.srcdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-codefirst",
            "src",
          ),
        });
        new SampleDir(this, this.testdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-codefirst",
            "test",
          ),
        });
      } else if (mergedOptions.sampleType === "graphql-schemafirst") {
        // generate-typings.ts uses ts-morph, so we need to add it to devDeps
        this.addDevDeps("ts-morph");
        const generateTypingsTs = "generate-typings.ts";

        for (const fileName of [generateTypingsTs, "nest-cli.json"]) {
          new SampleFile(this, fileName, {
            sourcePath: path.join(
              __dirname,
              "..",
              "samples",
              "nestjs-graphql-schemafirst",
              "root",
              fileName,
            ),
          });
        }
        new SampleDir(this, this.srcdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-schemafirst",
            "src",
          ),
        });
        new SampleDir(this, this.testdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-schemafirst",
            "test",
          ),
        });

        this.eslint?.addLintPattern(generateTypingsTs);
        this.tsconfigDev.addInclude(generateTypingsTs);
        this.addTask("generate-typings", {
          exec: `ts-node --project=tsconfig.dev.json ${generateTypingsTs}`,
          description: "Generate typings from graphql files",
        });
      } else {
        throw new Error("Unsupported sample type");
      }
    }
  }
}
