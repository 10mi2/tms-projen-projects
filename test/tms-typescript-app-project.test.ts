import { Testing } from "projen/lib/testing";
import { TmsTypeScriptAppProject } from "../src/";

test("TMSTypeScriptAppProject has reasonable configuration", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["tsconfig.json"]).toMatchSnapshot();
  expect(snapshot["tsconfig.dev.json"]).toMatchSnapshot();
  expect(snapshot["package.json"]).toMatchSnapshot();
  expect(snapshot[".vscode/settings.json"]).toMatchSnapshot();
  expect(snapshot[".vscode/extensions.json"]).toMatchSnapshot();

  expect(snapshot[".eslintrc.json"]).toMatchSnapshot();

  expect(snapshot[".nvmrc"]).toMatchSnapshot();

  expect(snapshot["src/index.ts"]).toBeDefined();
  expect(snapshot["src/hello.ts"]).toBeDefined();
  expect(snapshot["test/hello.test.ts"]).toBeDefined();

  const tasks = snapshot[".projen/tasks.json"].tasks;
  const bundleTask = tasks["bundle:index"];
  const bundleCommand = bundleTask.steps[0].exec;
  expect(bundleCommand).toContain("--format=esm");
  expect(bundleCommand).toContain("--banner:js=");
});

test("TMSTypeScriptAppProject doesn't make sample code when told not to", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    sampleCode: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["src/index.ts"]).toBeUndefined();
  expect(snapshot["src/hello.ts"]).toBeUndefined();
  expect(snapshot["test/hello.test.ts"]).toBeUndefined();
});

test("TMSTypeScriptAppProject doesn't make vscode settings when told not to", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    vscode: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot[".vscode/settings.json"]).toBeUndefined();
  expect(snapshot[".vscode/extensions.json"]).toBeUndefined();
});

test("TMSTypeScriptAppProject honors eslintFixableAsWarn=false", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot[".eslintrc.json"]).toMatchSnapshot(".eslintrc.json");
});

test("TMSTypeScriptAppProject honors esmSupportConfig=false", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    esmSupportConfig: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["package.json"].type).toBeUndefined();
  expect(snapshot["tsconfig.json"]).toMatchSnapshot("tsconfig.json");
  expect(snapshot["tsconfig.dev.json"]).toMatchSnapshot("tsconfig.dev.json");

  const tasks = snapshot[".projen/tasks.json"].tasks;
  const bundleTask = tasks["bundle:index"];
  const bundleCommand = bundleTask.steps[0].exec;
  expect(bundleCommand).toContain("--format=cjs");
});

test.each([true, false])(
  "TMSTypeScriptAppProject honors tsNodeUnknownFileExtensionWorkaround=%p",
  (tsNodeUnknownFileExtensionWorkaround: boolean) => {
    const project = new TmsTypeScriptAppProject({
      name: "test",
      defaultReleaseBranch: "main",
      eslintFixableAsWarn: false,
      esmSupportConfig: false,
      // default settings
      tsNodeUnknownFileExtensionWorkaround,
    });
    const snapshot = Testing.synth(project);

    const tasks = snapshot[".projen/tasks.json"].tasks;
    const defaultTask = tasks.default;
    const defaultCommand = defaultTask.steps[0].exec;
    if (tsNodeUnknownFileExtensionWorkaround) {
      expect(defaultCommand).toContain("--loader ts-node/esm");
    } else {
      expect(defaultCommand).not.toContain("--loader ts-node/esm");
    }
  },
);
describe.each([
  { version: "16.17.1", isOver18d19: false },
  { version: "16.20.2", isOver18d19: false },
  { version: "18.15.0", isOver18d19: false },
  { version: "18.17.0", isOver18d19: false },
  { version: "18.17.1", isOver18d19: false },
  { version: "18.18.0", isOver18d19: false },
  { version: "18.18.2", isOver18d19: false },
  { version: "18.19.0", isOver18d19: true },
  { version: "20.9.0", isOver18d19: true },
  { version: "20.10.0", isOver18d19: true },
])(
  "TMSTypeScriptAppProject interprets process.version=$version as >= 18.19.x: $isOver18d19",
  ({ version: nodeVersion, isOver18d19 }) => {
    const originalProcess = process;
    beforeEach(() => {
      global.process = {
        ...originalProcess,
        versions: { ...originalProcess.versions, node: nodeVersion },
      };
    });
    afterEach(() => {
      global.process = originalProcess;
    });

    test(`TMSTypeScriptAppProject interprets process.version=${nodeVersion} as ${isOver18d19 ? ">=" : "<"} 18.19.x`, () => {
      const project = new TmsTypeScriptAppProject({
        name: "test",
        defaultReleaseBranch: "main",
        eslintFixableAsWarn: false,
        esmSupportConfig: false,
        // default settings
      });
      const snapshot = Testing.synth(project);

      const tasks = snapshot[".projen/tasks.json"].tasks;
      const defaultTask = tasks.default;
      const defaultCommand = defaultTask.steps[0].exec;
      if (isOver18d19) {
        expect(defaultCommand).toContain("--loader ts-node/esm");
      } else {
        expect(defaultCommand).not.toContain("--loader ts-node/esm");
      }
    });
  },
);

test("TMSTypeScriptAppProject honors esmSupportConfig=false", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    esmSupportAddRequireShim: false,
    // default settings
    // esmSupportConfig: true,
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  const tasks = snapshot[".projen/tasks.json"].tasks;
  const bundleTask = tasks["bundle:index"];
  const bundleCommand = bundleTask.steps[0].exec;
  expect(bundleCommand).toContain("--format=esm");
  expect(bundleCommand).not.toContain("--banner:js=");
});

test("TMSTypeScriptAppProject honors strictest with workaround", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    esmSupportConfig: false,
    tsconfigBaseStrictest: true,
    tsconfigBaseNoArrayWorkaround: true,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["tsconfig.json"].extends).toEqual(
    "@tsconfig/node18/tsconfig.json",
  );
  expect(snapshot["tsconfig.json"].compilerOptions).toMatchObject({
    strict: true,
    exactOptionalPropertyTypes: true,
    noFallthroughCasesInSwitch: true,
    noImplicitOverride: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    noUncheckedIndexedAccess: true,
    noUnusedLocals: true,
    noUnusedParameters: true,

    isolatedModules: true,

    esModuleInterop: true,
    skipLibCheck: true,
    forceConsistentCasingInFileNames: true,

    allowUnusedLabels: false,
    allowUnreachableCode: false,
    checkJs: true,
  });
});

test("TMSTypeScriptAppProject honors strictest WITHOUT workaround", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    esmSupportConfig: false,
    tsconfigBaseStrictest: true,
    tsconfigBaseNoArrayWorkaround: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["tsconfig.json"].extends).toEqual([
    "@tsconfig/strictest/tsconfig.json",
    "@tsconfig/node18/tsconfig.json",
  ]);
});

test("TMSTypeScriptAppProject honors NOT strictest", () => {
  const project = new TmsTypeScriptAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    eslintFixableAsWarn: false,
    esmSupportConfig: false,
    tsconfigBaseStrictest: false,
    tsconfigBaseNoArrayWorkaround: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["tsconfig.json"].extends).toEqual(
    "@tsconfig/node18/tsconfig.json",
  );
});
