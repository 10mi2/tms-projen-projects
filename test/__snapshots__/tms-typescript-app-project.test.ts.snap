// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TMSTypeScriptAppProject has reasonable configuration 1`] = `
{
  "compilerOptions": {
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "checkJs": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "moduleResolution": "node16",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "lib",
    "rootDir": "src",
    "skipLibCheck": true,
    "strict": true,
  },
  "exclude": [],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
  ],
  "ts-node": {
    "esm": true,
    "experimentalSpecifierResolution": "node",
    "preferTsExts": false,
  },
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 2`] = `
{
  "compilerOptions": {
    "moduleResolution": "node16",
    "noEmit": true,
  },
  "exclude": [
    "node_modules",
  ],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
    "test/**/*.ts",
    ".projenrc.ts",
    "projenrc/**/*.ts",
  ],
  "ts-node": {
    "esm": true,
    "experimentalSpecifierResolution": "node",
    "preferTsExts": false,
  },
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 3`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
  "devDependencies": {
    "@tsconfig/node18": "*",
    "@types/jest": "*",
    "@types/node": "*",
    "@typescript-eslint/eslint-plugin": "^8",
    "@typescript-eslint/parser": "^8",
    "constructs": "^10.0.0",
    "esbuild": "*",
    "eslint": "^9",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "*",
    "jest-junit": "^16",
    "prettier": "*",
    "projen": ">=0.79.23",
    "ts-jest": "*",
    "ts-node": ">=10.9.2",
    "typescript": "*",
  },
  "jest": {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/",
    ],
    "coverageProvider": "v8",
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "moduleNameMapper": {
      "^(\\.{1,2}/.*)\\.m?js$": "$1",
    },
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": [
      "<rootDir>/@(src|test)/**/*(*.)@(spec|test).ts?(x)",
      "<rootDir>/@(src|test)/**/__tests__/**/*.ts?(x)",
      "<rootDir>/@(projenrc)/**/*(*.)@(spec|test).ts?(x)",
      "<rootDir>/@(projenrc)/**/__tests__/**/*.ts?(x)",
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
    ],
    "transform": {
      "^.+\\.m?[tj]sx?$": [
        "ts-jest",
        {
          "tsconfig": "tsconfig.dev.json",
          "useESM": true,
        },
      ],
    },
    "watchPathIgnorePatterns": [
      "/node_modules/",
    ],
  },
  "license": "Apache-2.0",
  "name": "test",
  "publishConfig": {
    "access": "public",
  },
  "scripts": {
    "build": "npx projen build",
    "bundle": "npx projen bundle",
    "bundle:index": "npx projen bundle:index",
    "bundle:index:watch": "npx projen bundle:index:watch",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "test": "npx projen test",
    "test:update": "npx projen test:update",
    "test:watch": "npx projen test:watch",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "type": "module",
  "version": "0.0.0",
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 4`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
  "jest.jestCommandLine": "npm test --",
  "jest.rootPath": "./",
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 5`] = `
{
  "recommendations": [
    "dbaeumer.vscode-eslint",
    "Orta.vscode-jest",
  ],
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 6`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "!.projenrc.ts",
    "!projenrc/**/*.ts",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.ts",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.dev.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "dot-notation": [
      "warn",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/test/**",
          "**/build-tools/**",
          ".projenrc.ts",
          "projenrc/**/*.ts",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "warn",
    ],
    "no-bitwise": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "warn",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "warn",
    ],
    "prettier/prettier": [
      "warn",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.dev.json",
      },
    },
  },
}
`;

exports[`TMSTypeScriptAppProject has reasonable configuration 7`] = `"v18.18.2"`;

exports[`TMSTypeScriptAppProject honors eslintFixableAsWarn=false: .eslintrc.json 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "!.projenrc.ts",
    "!projenrc/**/*.ts",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.ts",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.dev.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/test/**",
          "**/build-tools/**",
          ".projenrc.ts",
          "projenrc/**/*.ts",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "error",
    ],
    "no-bitwise": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.dev.json",
      },
    },
  },
}
`;

exports[`TMSTypeScriptAppProject honors esmSupportConfig=false: tsconfig.dev.json 1`] = `
{
  "compilerOptions": {
    "moduleResolution": "node16",
    "noEmit": true,
  },
  "exclude": [
    "node_modules",
  ],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
    "test/**/*.ts",
    ".projenrc.ts",
    "projenrc/**/*.ts",
  ],
  "ts-node": {
    "compilerOptions": {
      "module": "commonjs",
    },
    "experimentalSpecifierResolution": "node",
    "preferTsExts": true,
  },
}
`;

exports[`TMSTypeScriptAppProject honors esmSupportConfig=false: tsconfig.json 1`] = `
{
  "compilerOptions": {
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "checkJs": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "moduleResolution": "node16",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "lib",
    "rootDir": "src",
    "skipLibCheck": true,
    "strict": true,
  },
  "exclude": [],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
  ],
  "ts-node": {
    "compilerOptions": {
      "module": "commonjs",
    },
    "experimentalSpecifierResolution": "node",
    "preferTsExts": true,
  },
}
`;
